### PR TITLE
 [MIG] stock_available_mrp: Migration to 11.0 

### PR DESCRIPTION
**Pull request by tschanzt**

### CODE REVIEW

File: stock_available_mrp/models/product_product.py
Comment (grindtildeath) [inline stock_available_mrp/models/product_product.py]: to respect the inheritance, it would be better to call super here, and change the results in your for loop, because other modules might be using the same function to compute some other fields, and the inheritance would be broken.

File: stock_available_mrp/models/product_product.py
Comment (grindtildeath) [inline stock_available_mrp/models/product_product.py]: IMO this should be `@api.multi` without the product arg.
You can add `self.ensure_one()` if it's meant to be called only for one product.

File: stock_available_mrp/tests/test_potential_qty.py
Comment (grindtildeath) [inline stock_available_mrp/tests/test_potential_qty.py]: If he can't compute the potential, shouldn't we have `0` here ? :confused:

File: stock_available_mrp/tests/test_potential_qty.py
Comment (grindtildeath) [inline stock_available_mrp/tests/test_potential_qty.py:260]: Can you explain these change of 42 to 313 ? If the quantities don't change in the setup, I find it weird it's changed in the result of the test.

File: stock_available_mrp/models/product_template.py
Comment (grindtildeath) [inline stock_available_mrp/models/product_template.py]: Same as on `product.product`, please respect the inheritance and call super().

**Review by grindtildeath — CHANGES_REQUESTED** [review]
Please attend these comments and fix the error encountered on runbot trying to create a new product (might need a new PR to fix stock_available though) : 
```
Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/fields.py", line 937, in __get__
    value = record.env.cache.get(record, self)
  File "/.repo_requirements/odoo/odoo/api.py", line 960, in get
    value = self._data[field][record.id][key]
KeyError: <odoo.api.Environment object at 0x7f8a9a05ce80>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/http.py", line 650, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/.repo_requirements/odoo/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/.repo_requirements/odoo/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/.repo_requirements/odoo/odoo/http.py", line 692, in dispatch
    result = self._call_function(**self.params)
  File "/.repo_requirements/odoo/odoo/http.py", line 342, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/http.py", line 335, in checked_call
    result = self.endpoint(*a, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 936, in __call__
    return self.method(*args, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/OCB-11.0/addons/web/controllers/main.py", line 931, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/home/odoo/OCB-11.0/addons/web/controllers/main.py", line 923, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 689, in call_kw
    return call_kw_multi(method, model, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 680, in call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/models.py", line 5091, in onchange
    record.mapped(dotname)
  File "/.repo_requirements/odoo/odoo/models.py", line 4504, in mapped
    recs = recs._mapped_func(operator.itemgetter(name))
  File "/.repo_requirements/odoo/odoo/models.py", line 4483, in _mapped_func
    vals = [func(rec) for rec in self]
  File "/.repo_requirements/odoo/odoo/models.py", line 4483, in <listcomp>
    vals = [func(rec) for rec in self]
  File "/.repo_requirements/odoo/odoo/models.py", line 4750, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/.repo_requirements/odoo/odoo/fields.py", line 943, in __get__
    self.determine_draft_value(record)
  File "/.repo_requirements/odoo/odoo/fields.py", line 1063, in determine_draft_value
    self._compute_value(record)
  File "/.repo_requirements/odoo/odoo/fields.py", line 999, in _compute_value
    getattr(records, self.compute)()
  File "/home/odoo/build/OCA/stock-logistics-warehouse/stock_available/models/product_template.py", line 28, in _compute_available_quantities
    for key, value in res[product.id].items():
KeyError: <odoo.models.NewId object at 0x7f8a9aab7ac8>
```

File: stock_available_mrp/tests/test_potential_qty.py
Comment (tschanzt) [reply stock_available_mrp/tests/test_potential_qty.py]: @grindtildeath He expects an error to raise in _get_components_needs which is caught and sets the value to 0 but in v11 the error didn't get raised anymore. I don't think that i removed any conditions, but there were some changes to mrp. We can add a manual check or maybe we can get rid of the test.

File: stock_available_mrp/tests/test_potential_qty.py
Comment (tschanzt) [reply stock_available_mrp/tests/test_potential_qty.py:260]: @grindtildeath The Tests depended on Demodata of product, mrp. etc which has changed or was removed since v9. So i tried to recreate the data. Here there probably is a product_qty missing. Since its Likely that the original bom line had a efficiency on it, I ditdn't try to mach it exactly.

File: stock_available_mrp/README.rst
Comment (simahawk) [inline stock_available_mrp/README.rst:21]: :fire:

File: stock_available_mrp/README.rst
Comment (simahawk) [inline stock_available_mrp/README.rst:58]: :fire:

File: stock_available_mrp/__manifest__.py
Comment (simahawk) [inline stock_available_mrp/__manifest__.py:9]: `'website': 'https://github.com/OCA/stock-logistics-warehouse',`

File: stock_available_mrp/demo/mrp_data.xml
Comment (simahawk) [inline stock_available_mrp/demo/mrp_data.xml]: `noupdate` can stay on `odoo` tag and you can remove `data`

File: stock_available_mrp/models/product_product.py
Comment (simahawk) [inline stock_available_mrp/models/product_product.py]: too many empty lines

File: stock_available_mrp/models/product_product.py
Comment (simahawk) [inline stock_available_mrp/models/product_product.py]: `_compute_component_ids`

File: stock_available_mrp/models/product_product.py
Comment (simahawk) [inline stock_available_mrp/models/product_product.py:67]: too many empty lines

File: stock_available_mrp/models/product_product.py
Comment (simahawk) [inline stock_available_mrp/models/product_product.py]: should be @api.multi + loop and collect all the product ids and browse only once at the end.

File: stock_available_mrp/tests/test_potential_qty.py
Comment (simahawk) [inline stock_available_mrp/tests/test_potential_qty.py]: tracking_disable, no_reset_password

File: stock_available_mrp/tests/test_potential_qty.py
Comment (simahawk) [inline stock_available_mrp/tests/test_potential_qty.py]: too many empty lines

File: stock_available_mrp/tests/test_potential_qty.py
Comment (simahawk) [inline stock_available_mrp/tests/test_potential_qty.py:14]: tracking_disable

**Review by simahawk — CHANGES_REQUESTED** [review]

File: stock_available_mrp/models/product_product.py
Comment (tschanzt) [reply stock_available_mrp/models/product_product.py]: I think we can drop this compute function since it is currently never called since component_ids is never accessed directly. component_ids was used for an api.depends so we would update the quantity if it changed on one of the components. But since the potential_qty isn't stored we would recalculate it anyways if accessed. Do you agree or am i missing something? /cc @grindtildeath

Comment (grindtildeath) [reply stock_available_mrp/models/product_product.py]: @tschanzt Seems to make sense although I fear performance could be an issue not storing `potential_qty`...
Let us have a look together on monday :wink:

**Comment by tschanzt** [comment]
@grindtildeath I couldn't reproduce the error locally and with the new runbot build it seems to work.

File: stock_available_mrp/models/product_product.py
Comment (grindtildeath) [reply stock_available_mrp/models/product_product.py]: @tschanzt ~Allright let's remove this function and `_compute_available_quantities` and we'll be good !~ :+1:

Comment (grindtildeath) [reply stock_available_mrp/models/product_product.py]: @tschanzt As we've seen together, we need a relation to `potential_qty` in the `api.depends` of the compute function to ensure it's calculated correctly as recursive. So I suppose we still need to compute the components of each product to use this relation and ensure the components are computed in the right order.
IMO would be great to add a second test function based on `test_potential_qty__list` but with the creation of the components and finished product in the reverse order, to ensure we don't get the right result by a random order in the test.

File: stock_available_mrp/models/product_product.py
Comment (grindtildeath) [inline stock_available_mrp/models/product_product.py]: Should be returned and not assigned to potential_qty.

**Review by grindtildeath — CHANGES_REQUESTED** [review]
I will open a PR on your branch to fix this :wink:

**Comment by rafaelbn** [comment]
Hi @tschanzt ,

I think you should start migration from https://github.com/OCA/stock-logistics-warehouse/pull/280

Don you think so?

:thinking:

**Comment by grindtildeath** [comment]
Hi @rafaelbn ,

Well, it would have been done, but when @tschanzt opened this PR, the PR you mentioned was not working and stalled without activity for months although reviews were done for months as well.
So as we didn't have the need to migrate this module on v10, we migrated it straight from v9.

**Comment by rafaelbn** [comment]
Thanks @grindtildeath . I'm just trying to help. It looks that https://github.com/OCA/stock-logistics-warehouse/pull/280  should be merged ( @Cedric-Pigeon is not sure about why travis is :red_circle: ) by any @OCA/logistics-maintainers . Also @Cedric-Pigeon could be interested in reviewing this PR :smile: 

I can make a functional review here once is fixed  :smiley:

**Comment by rafaelbn** [comment]
Nice functionally. @simahawk @grindtildeath maybe you would like to take a look again :smile:

**Review by rafaelbn — APPROVED** [review]
Tested in runbot :+1:

File: stock_available_mrp/models/product_product.py
Comment (simahawk) [inline stock_available_mrp/models/product_product.py]: indent this comment

File: stock_available_mrp/README.rst
Comment (simahawk) [inline stock_available_mrp/README.rst:58]: we don't have this link anymore as it's hard to maintain https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

File: stock_available_mrp/models/product_template.py
Comment (simahawk) [inline stock_available_mrp/models/product_template.py]: s/iincludes/includes

**Review by simahawk — APPROVED** [review]
still some minor glitches. Approving to not block when ready ;)

**Comment by tschanzt** [comment]
@simahawk i fixed the mentioned issues.

**Review by grindtildeath — APPROVED** [review]

**Comment by simahawk** [comment]
@tschanzt good. Squash your commits pls, then we are ready to merge when it gets green ;)

**Comment by tschanzt** [comment]
@simahawk squashed and its green.

**Comment by lmignon** [comment]
@tschanzt #280 includes a lot of improvement especially in terms of performance.. I think it would be necessary to complete the review of #280 and include its improvements into the current one.

**Review by lmignon — CHANGES_REQUESTED** [review]
@tschanzt #280 includes a lot of improvement especially in terms of performance.. I think it would be necessary to complete the review of #280 and include its improvements into the current one.

**Comment by rousseldenis** [comment]
@tschanzt Can you attend @lmignon comments ?

**Comment by tschanzt** [comment]
@rousseldenis yeah sure.

**Comment by tschanzt** [comment]
@rousseldenis Done. Its based on #280 now. Could you take a look and Review?

**Comment by rafaelbn** [comment]
@tschanzt check why Travis is :red_circle:

**Comment by rousseldenis** [comment]
@lmignon Could you update your review ?

**Comment by rafaelbn** [comment]
I think this PR should be merged :+1: don't you think so @lmignon ?

**Comment by rafaelbn** [comment]
I'm not PSC but I think we could merge as https://github.com/OCA/stock-logistics-warehouse/pull/280

cc @OCA/logistics-maintainers

**Review by rafaelbn — APPROVED** [review]
:+1:

**Comment by rousseldenis** [comment]
@tschanzt You will squash last two commits or let it as is ?

**Comment by tschanzt** [comment]
@rousseldenis the last two commits are squashed now.

File: stock_available_mrp/models/product_product.py
Comment (rousseldenis) [inline stock_available_mrp/models/product_product.py:118]: @tschanzt Just a question about this. The standard option is 'UP'. So, why rounding down as it could be weird for users if behaviour is not the same on other places?

Comment (tschanzt) [reply stock_available_mrp/models/product_product.py:118]: @rousseldenis The point of it is to be pesimistic about how much you can deliver. The original problem was, that we didn't respect the rounding of the uom. E.G. if you want to make a skateboard (requiring the board and 4 wheels) and you had the board and 2 wheels it would return 0.5 instead of 0 so i added this to apply the rounding, but if we round up we then get 1 as the potential quantity which isn't what we can produce.

Comment (rousseldenis) [reply stock_available_mrp/models/product_product.py:118]: Makes sense ! So, let's go

**Pull request merged**
